core/federated/RTI/rti_remote.c

```diff
@@ -1063,32 +1064,52 @@ void handle_address_ad(uint16_t federate_id) {
 }
 
 /**
- * Send to the start time to the federate my_fed.
- * This function assumes the caller does not hold the mutex.
+ * Send the start time and tag to the federate my_fed.
  *
- * If it is the startup phase, the start_time will be the maximum received timestamps
- * plus an offset. The federate will then receive identical federation_start_time
- * and federate_start_tag.time (the federate_start_tag.microstep will be 0).
- * If, however, the startup phase is passed, the federate will receive different
- * values than stated above.
+ * During the startup phase, the start_time is calculated as the maximum received timestamp, plus an offset.
+ * The federate will then receive identical values for federation_start_time` and `federate_start_tag.time` (with
+ * `federate_start_tag.microstep` set to 0). After the startup phase, the federate will receive different values for
+ * these parameters.
  *
- * This will also notify federates downstream of my_fed that this federate is now
- * connected.  This is important when there are zero-delay cycles.
+ * Before sending the start time and tag, this function performs the following actions:
+ *  - If my_fed is transient, notify federates downstream of its connection, ensuring proper handling of zero-delay
+ *    cycles.
+ *  - Notify my_fed of all upstream transient federates that are connected.
  *
- * This function assumes the caller holds the mutex.
+ * This function assumes that the mutex lock is already held.
  *
  * @param my_fed the federate to send the start time to.
  * @param federation_start_time the federation start_time
  * @param federate_start_tag the federate effective start tag
  */
-static void send_start_tag(federate_info_t* my_fed, instant_t federation_start_time, tag_t federate_start_tag) {
+static void send_start_tag_locked(federate_info_t* my_fed, instant_t federation_start_time, tag_t federate_start_tag) {
+  // If this is a transient federate, notify its downstream federates that it is now connected.
+  if (my_fed->is_transient) {
+    for (int i = 0; i < my_fed->enclave.num_downstream; i++) {
+      send_upstream_connected_locked(GET_FED_INFO(my_fed->enclave.downstream[i]), my_fed);
+    }
+  }
+  // A corner case occurs when an upstream transient joins at tag (0, 0), but `my_fed`
+  // either misses the notification or receives it late. The following ensures that
+  // `my_fed` is informed of all currently connected upstream transients.
+  // This also prevents `my_fed` from receiving the start time and starting execution
+  // before the upstream connection message is received.
+  // This also deals with an even less likely corner case where two transients are joining simultaneously and one is
+  // upstream of the other.
```

Comment:
I would simplify this comment. It's not really a corner case. Suggestion:
```
  // Notify my_fed of any upstream transient federates that are connected.
  // This has to occur before sending the start tag so that my_fed does not begin executing thinking that these
  // upstream federates are not connected.
```

Reply:
Added to the upcoming commit.